src/index.js

```diff
@@ -243,6 +246,45 @@ module.exports = (web3, options = {}) => {
      * @return {Promise} A promise that resolves to a raw transaction
      */
     async publishVersion (manager, appId, version, provider, directory, contract, from) {
+      const {
+        targetContract,
+        name,
+        params
+      } = await this.publishVersionIntent(manager, appId, version, provider, directory, contract)
+
+      try {
+        const call = targetContract.methods[name](...params)
+
+        // Return transaction to sign
+        return {
+          to: targetContract.options.address,
+          data: call.encodeABI(),
+          gas: Math.round(await call.estimateGas({ from }) * GAS_FUZZ_FACTOR),
+          gasPrice: web3.utils.toWei('10', 'gwei'),
+          nonce: await web3.eth.getTransactionCount(manager)
+        }
+      } catch (err) {
+        throw new Error(`Transaction would not succeed ("${err.message}")`)
+      }
+    },
+
+    /**
+     * Publishes a new version (`version`) of `appId` using storage provider `provider`.
+     *
+     * If the destination repository does not exist, the intent will be for creating a new
+     * repository with an initial version.

```

Comment:
Is it a lot easier for the CLI to use a single function to do both version publishes as well as repo creations?

Really seems like there's two concerns mixed here (you can immedately tell by the `manager` being the first argument, but it not being used at all on the "normal" path).

Reply:
Not necessarily, I think having something like:

```
if ( apm.getRepo() ) {
  apm.publishVersion()
} else {
  apm.publishInitialVersion()
}
```

is not more difficult, but actually more expressive :+1: (you don't have to guess the magic `publishVersion` does).

That being said, refactoring this might take some time without adding a lot of value (especially if we want to keep backwards compatibility), so I will open an issue for it.